runtime/src/qio/qio_formatted.c

```diff
@@ -361,6 +361,84 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
   return 0;
 }
 
+qioerr read_entire_file(const int threadsafe,qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{  
+  qioerr err,err_reading_file=0;
+  int64_t file_length;
+  char* ret=NULL;
+  ssize_t amt=0;
+  err_t errcode;
+  int64_t fileLenGuess=1024;
+  ssize_t total_amt_read=0;
+  bool err_flag;
+  if( threadsafe ) 
+  {
+    err = qio_lock(&ch->lock);
+    if( err ) return err;
+  }
+
+  err=qio_file_length(ch->file,&file_length);   // obtain file length
+ 
+  if(!err)                                      // no error in obtaining file length
+  {
+   
+    ret=qio_malloc(file_length+1);              //allocate initial buffer
+    ret[0]='\0';
+    err=qio_channel_read(false, ch, ret,file_length+1, &amt);    //file_length+1 instead of file_length -->I try to read one more character than the prescribed 
+                                                                 //file length . That way, if we have already reached the file end, reading 1 more character will 
+                                                                 // will generate EOF. But, if the file length has increased while reading, we have just read one more
+                                                                 // character.
+    if(qio_err_to_int(err)==EEOF)                                // Here,checking whether file end is reached.
+    {
+       err=0;
+    }
+    else
+    {
+       err=QIO_EEOF;
+    }
+    total_amt_read=amt;
+  }
+
+  if(err)                                                          // error might be due to being unable to fetch file length or there is something left to read.
+  {    
+    ret=qio_realloc(ret,total_amt_read+fileLenGuess+1);            // add some more space to the buffer to read the remaining data.
+    while(true)
+    {
+      err_reading_file = qio_channel_read(false, ch,&ret[total_amt_read],1024LL, &amt);   // we will read the remaining data in units of 1024 each time .
+      total_amt_read+=amt;                                                                //calculate the amt that we have read until now
+      if (err_reading_file == QIO_EEOF) break;                                            // break out of loop is data is exhausted finally
+      fileLenGuess=total_amt_read+1024;
+      ret = qio_realloc(ret, fileLenGuess+1);
+    }
+    file_length=total_amt_read;                                                            //get the new file length
```

Comment:
I think the code will be clearer if, in this line and the rest of the function, you just use `total_amt_read` instead of the `file_length` variable. (So, remove line 413, use total_amt_read in 414 and 419). That way, file_length is only relevant for the first part of the function.